### DISCUSSION TITLE
Using ADT -installAPP after having successfully created AAB file.

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#1078] [Q&A] [answered]
After having successfully created a .aab file for my sample app with a lot of help from @ajwfrost , I am trying use ADT's installApp to try and install the app on my test device.

Below is the log of aab file creation:

`*** ADT Logging at 2021/08/16 18:48:46 ***`
`Called by: D:\__SDK\AIR-sdk\lib\adt.jar`
`ADT 33.1.1.575 called with: -package -target aab -storetype pkcs12 -keystore ..\__certificate\MediaproCertificate.p12 -storepass xxxxxx C:\Users\MJP\Desktop\UISampler.aab application.xml -C bin . -C icons . -extdir ..\__extensions -platformsdk D:\__SDK\android-sdk`
`AAB Packager entry point`
`Version in APK = 33.1.1.575, version of ADT = 33.1.1.575`
`Version in APK = 33.1.1.575, version of ADT = 33.1.1.575`
`Version in APK = 33.1.1.575, version of ADT = 33.1.1.575`
`Version in APK = 33.1.1.575, version of ADT = 33.1.1.575`
`Gradle build command:  D:\__SDK\openjdk8\bin\java.exe -Xmx1024m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx1024m -Duser.dir="C:\Users\MJP\AppData\Local\Temp\3af4c735-ad6e-4749-8ecf-dd8d3b8cc0d9" -cp "D:\__SDK\AIR-sdk\lib\android\bin\gradle-wrapper.jar";"D:\__SDK\AIR-sdk\lib\android\bin\gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:bundleRelease`

aab file created successfully.

Below is a log of failure to install the generated aab on the test device.

`*** ADT Logging at 2021/08/16 18:50:24 ***
`Called by: D:\__SDK\AIR-sdk\lib\adt.jar
`ADT 33.1.1.575 called with: -installApp -platform android -platformsdk D:\__SDK\android-sdk -device 3fa1f4d6 -package UISampler.aab`
`Running Java command line: java -jar D:\__SDK\AIR-sdk\lib\android\lib\bundletool.jar build-apks --bundle=C:\Users\MJP\Desktop\UISampler.aab --output=C:\Users\MJP\AppData\Local\Temp\aab8158506459675556279.apks --overwrite`
`Running Java command line: java -jar D:\__SDK\AIR-sdk\lib\android\lib\bundletool.jar install-apks --apks=C:\Users\MJP\AppData\Local\Temp\aab8158506459675556279.apks --device-id=3fa1f4d6 --adb=D:\__SDK\android-sdk\platform-tools\adb.exe`
`The APKs have been extracted in the directory: C:\Users\MJP\AppData\Local\Temp\3387726573426748004`
`[BT:0.10.3] Error: Installation of the app failed.`
`com.android.tools.build.bundletool.model.exceptions.InstallationException: Installation of the app failed.`
`	at com.android.tools.build.bundletool.model.exceptions.InstallationException$Builder.build(InstallationException.java:47)
	at com.android.tools.build.bundletool.model.exceptions.InstallationException$Builder.build(InstallationException.java:41)
	at com.android.tools.build.bundletool.device.DdmlibDevice.installApks(DdmlibDevice.java:124)
	at com.android.tools.build.bundletool.device.ApksInstaller.installOnDevice(ApksInstaller.java:94)
	at com.android.tools.build.bundletool.device.ApksInstaller.installApks(ApksInstaller.java:82)
	at com.android.tools.build.bundletool.device.ApksInstaller.installApks(ApksInstaller.java:62)
	at com.android.tools.build.bundletool.commands.InstallApksCommand.execute(InstallApksCommand.java:163)
	at com.android.tools.build.bundletool.BundleToolMain.main(BundleToolMain.java:88)
	at com.android.tools.build.bundletool.BundleToolMain.main(BundleToolMain.java:46)`
`Caused by: com.android.ddmlib.InstallException: Failed to finalize session : INSTALL_PARSE_FAILED_NO_CERTIFICATES: Package /data/app/vmdl449741274.tmp/0_base-xhdpi has no certificates at entry AndroidManifest.xml`
`	at com.android.ddmlib.SplitApkInstaller.install(SplitApkInstaller.java:91)`
`	at com.android.ddmlib.Device.installPackages(Device.java:928)`
`	at com.android.tools.build.bundletool.device.DdmlibDevice.installApks(DdmlibDevice.java:108)`
`	... 6 more`
`..java.lang.Thread.getStackTrace(Unknown Source)`
`..com.adobe.air.DebugFile.printStack(DebugFile.java)`
`..com.adobe.air.ADTException.<init>(ADTException.java)`
`..com.adobe.air.android.AndroidDeviceOperations.installAAB(AndroidDeviceOperations.java)`
`..com.adobe.air.android.AndroidDeviceOperations.installApplication(AndroidDeviceOperations.java)`
`..com.adobe.air.ADT.parsePlatformOperationArgs(ADT.java)`
`..com.adobe.air.ADT.parseArgsAndGo(ADT.java)`
`..com.adobe.air.ADT.run(ADT.java)`
`..com.adobe.air.ADT.main(ADT.java)`
`Stack trace:`
`ADT exception: Failed to install C:\Users\MJP\Desktop\UISampler.aab:`
`com.android.tools.build.bundletool.model.exceptions.InstallationException: Installation of the app failed.`
`Caused by: com.android.ddmlib.InstallException: Failed to finalize session : INSTALL_PARSE_FAILED_NO_CERTIFICATES: Package /data/app/vmdl449741274.tmp/0_base-xhdpi has no certificates at entry AndroidManifest.xml`

I tried this with `-target aab` and `-target aab-debug` and get the same result both.

Is there something else I should be doing?
Thanks.

**Comment by ajwfrost** [comment]
I believe someone else found the same issue, due to a missing Android debug certificate... see the below article under "Sign your debug build":
https://developer.android.com/studio/publish/app-signing
If you check for that debug.keystore, you may find it's missing: in which case, please use Android Studio and build/run a test app, and then it should be generated.

We could potentially add the requirement to pass in apk code-signing details to the installApp command, in order to avoid the dependency upon this debug keystore, but thought it was better to go with this practice (basically, it's what you get when using the bundletool if you don't specify your own certificate..)

thanks

**Reply by manishpatel0404** [reply]
Thanks @ajwfrost - Forgive me for being a little old-school, but in my entire journey, starting with compiling the AIR app, creating native and java-based ANEs, packaging and running APKs on emulators and devices, and finally creating the .aab file with your support, I have not had the need to install Android Studio or any other IDE, purely NotePad++ and command line. It would be silly to install it now, only to generate a debug.keystore (which, as you mentioned, is missing, probably because I have not installed Android Studio).

> We could potentially add the requirement to pass in apk code-signing details to the installApp command, in order to avoid the dependency upon this debug keystore

I would really appreciate this. I am not sure if other developers would agree, but I find it a far more convenient feature than to install Android Studio just to generate the debug.keystore. I am sure Android Studio would be useful in many aspects of development, but for pure-play AS3 apps which do not use too many device features, it seems like an overkill.

I would recommend the same for signing aab files before uploading the aab to the play store. This will complete the circle.

I am still reading up on the link you sent me, especially the "Generate an upload key and keystore" and "Sign your app with your upload key" sections of the link. I am not sure if these features are already available or can be added to ADT, maybe something like `adt -sign SIGNING_OPTIONS -target aab`

I have not had much experience with Android and have persevered to reach where I have. So please disregard anything that may sound outlandish.

I ended up trying the following:

**1) Generate debug.keystore** REF: [https://coderwall.com/p/r09hoq/android-generate-release-debug-keystores](url)
`keytool -genkey -v -keystore C:\Users\MJP\.android\debug.keystore -storepass android -alias androiddebugkey -keypass android -keyalg RSA -keysize 2048 -validity 10000`

**2) Generate the apks file**
`java -jar D:\__SDK\AIR-sdk\lib\android\lib\bundletool.jar build-apks --bundle UISampler.aab --output UISampler.apks --overwrite --ks=C:\Users\MJP\.android\debug.keystore --ks-pass=pass:android --ks-key-alias=androiddebugkey --key-pass=pass:android --connected-device`

**3) Install the apks to the test device**
`java -jar D:\__SDK\AIR-sdk\lib\android\lib\bundletool.jar install-apks --apks=UISampler.apks`

It seemed to install the apk on my test device. Surprisingly, the app's folder was not created in Android/data/. All I could see was the app's icon and when I tapped it, the screen went blank and returned without loading the main screen. What am I doing wrong here?

Thanks.

**Reply by ajwfrost** [reply]
Just regarding the idea of `adt -sign SIGNING_OPTIONS -target aab`:
- the AAB files can already be signed, they just take a jarsigner type format (like APK v1 signing) and this happens if you pass in the normal signing options to `-target aab` the same as you'd do for `-target apk`. 
- however, because these are completely different formats, the `bundletool` library needs to be given a new certificate that it can use for apk-signing (v2) the output APKs. So, we would actually need to set it up with something like:
`adt -installApp -platform android -storetype jks -keystore debug.keystore -storepass pass:android -alias androiddebugkey -keypass pass:android -package UISampler.aab`
... bit of a mouthful!

In terms of what's happening when you then start up the app... is there any information in the logcat output? A relatively common thing is if there is a missing ANE implementation for one of the platforms but if it was working with APK builds for that device, it should work for a Bundle that's then set up for the same device.

thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - please find below, the logcat output after having installed and run the app for the first time. I have no idea what all this means. I tested with an apk of the app and it works perfectly.

`08-17 13:13:03.777 I/Timeline( 1576): Timeline: Activity_launch_request id:air.com.mediapro.UISampler time:7378255`
`08-17 13:13:03.779 I/ActivityManager(  877): START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=air.com.mediapro.UISampler/.AIRAppEntry (has extras)} from uid 10060 on display 0`
`08-17 13:13:03.790 W/BroadcastQueue(  877): Permission Denial: receiving Intent { act=com.android.launcher3.action.LAUNCH flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$GmsExternalReceiver requires com.android.launcher3.permission.RECEIVE_LAUNCH_BROADCASTS due to sender com.android.launcher3 (uid 10060)`
`08-17 13:13:03.793 D/BubbleTextView( 1576): drawUnreadEvent() this = com.android.launcher3.BubbleTextView{2c00f25a VFED..CL ...P.... 501,412-668,618 #b}`
`08-17 13:13:03.833 I/ActivityManager(  877): Start proc air.com.mediapro.UISampler for activity air.com.mediapro.UISampler/.AIRAppEntry: pid=11403 uid=10097 gids={50097, 9997, 3003, 1028, 1015} abi=arm64-v8a`
`08-17 13:13:03.838 D/BubbleTextView( 1576): drawUnreadEvent() this = com.android.launcher3.BubbleTextView{2c00f25a VFED..CL ........ 501,412-668,618 #b}`
`08-17 13:13:03.843 D/BubbleTextView( 1576): drawUnreadEvent() this = com.android.launcher3.BubbleTextView{2c00f25a VFED..CL ........ 501,412-668,618 #b}`
`08-17 13:13:03.859 D/BubbleTextView( 1576): drawUnreadEvent() this = com.android.launcher3.BubbleTextView{2c00f25a VFED..CL ........ 501,412-668,618 #b}`
`08-17 13:13:03.871 D/PhoneStatusBar( 1131): setSystemUiVisibility vis=0 mask=ffffffff oldVal=8600 newVal=0 diff=8600`
`08-17 13:13:03.873 D/PhoneStatusBar( 1131): hiding the MENU button`
`08-17 13:13:03.891 W/ResourceType(11403): ResTable_typeSpec entry count inconsistent: given 10, previously 14`
`08-17 13:13:03.904 I/MultiDex(11403): VM with version 2.1.0 has multidex support`
`08-17 13:13:03.904 I/MultiDex(11403): Installing application`
`08-17 13:13:03.904 I/MultiDex(11403): VM has multidex support, MultiDex support library is disabled.`
`08-17 13:13:03.921 I/StartupTime1(11403): :1629213183921`
`08-17 13:13:03.921 I/AIR VERSION(11403): :33.1.1.575`
`08-17 13:13:03.922 I/DEVICE ARCH(11403): :arm64-v8a`
`08-17 13:13:04.086 D/PhoneStatusBar( 1131): setSystemUiVisibility vis=1707 mask=ffffffff oldVal=0 newVal=1707 diff=1707`
`08-17 13:13:04.091 D/PhoneStatusBar( 1131): hiding the MENU button`
`08-17 13:13:04.139 I/ServiceManager(11225): Waiting for service AtCmdFwd...`
`08-17 13:13:04.152 I/AIR RUNTIME(11403): Version 33,1,1,575 - platform Android-ARM64`
`08-17 13:13:04.625 D/PhoneStatusBar( 1131): setSystemUiVisibility vis=0 mask=ffffffff oldVal=1707 newVal=0 diff=1707`
`08-17 13:13:04.627 D/PhoneStatusBar( 1131): hiding the MENU button`
`08-17 13:13:04.925 I/WindowManager(  877): Screenshot max retries 4 of Token{1720c6d7 ActivityRecord{64c7656 u0 air.com.mediapro.UISampler/.AIRAppEntry t43 f}} appWin=Window{165b2563 u0 SurfaceView} drawState=3`
`08-17 13:13:04.925 E/ActivityManager(  877): Invalid thumbnail dimensions: 384x384`
`08-17 13:13:04.933 I/Choreographer(11403): Skipped 50 frames!  The application may be doing too much work on its main thread.`
`08-17 13:13:04.970 W/InputMethodManagerService(  877): Focus gain on non-focused client com.android.internal.view.IInputMethodClient$Stub$Proxy@348f1ed5 (uid=10097 pid=11403)`
`08-17 13:13:04.996 D/PhoneStatusBar( 1131): setSystemUiVisibility vis=600 mask=ffffffff oldVal=0 newVal=600 diff=600`
`08-17 13:13:04.996 D/PhoneStatusBar( 1131): hiding the MENU button`
`08-17 13:13:05.037 W/InputMethodManagerService(  877): Window already focused, ignoring focus gain of: com.android.internal.view.IInputMethodClient$Stub$Proxy@146fb942 attribute=null, token = android.os.BinderProxy@59e3c1a`
`08-17 13:13:05.037 I/Timeline( 1576): Timeline: Activity_idle id: android.os.BinderProxy@ee114ae time:7379515`
`08-17 13:13:05.042 D/PhoneStatusBar( 1131): setSystemUiVisibility vis=8600 mask=ffffffff oldVal=600 newVal=8600 diff=8000`
`08-17 13:13:05.043 D/PhoneStatusBar( 1131): hiding the MENU button`
`08-17 13:13:05.055 I/Process (11403): Sending signal. PID: 11403 SIG: 9`
`08-17 13:13:05.139 I/ServiceManager(11225): Waiting for service AtCmdFwd...`
`08-17 13:13:05.167 I/WindowState(  877): WIN DEATH: Window{19cc2fe1 u0 air.com.mediapro.UISampler/air.com.mediapro.UISampler.AIRAppEntry}`
`08-17 13:13:05.170 I/WindowState(  877): WIN DEATH: Window{165b2563 u0 SurfaceView EXITING}`
`08-17 13:13:05.173 I/WindowState(  877): WIN DEATH: Window{9102d53 u0 SurfaceView EXITING}`
`08-17 13:13:05.174 I/ActivityManager(  877): Process air.com.mediapro.UISampler (pid 11403) has died`
`08-17 13:13:05.293 I/Timeline(  877): Timeline: Activity_windows_visible id: ActivityRecord{950c8ec u0 com.android.launcher3/.Launcher t1} time:7379771`
`08-17 13:13:05.460 I/Timeline(  877): Timeline: Activity_windows_visible id: ActivityRecord{64c7656 u0 air.com.mediapro.UISampler/.AIRAppEntry t43 f} time:7379937`
`08-17 13:13:06.140 I/ServiceManager(11225): Waiting for service AtCmdFwd...`
`08-17 13:13:07.140 W/Atfwd_Sendcmd(11225): AtCmdFwd service not published, waiting... retryCnt : 2`
`08-17 13:13:07.674 E/QCOMSysDaemon(11441): Can't find/open bootselect node: (No such file or directory)`
`08-17 13:13:07.674 I/QCOMSysDaemon(11441): Starting qcom system daemon`
`08-17 13:13:07.674 E/Diag_Lib(11441):  Diag_LSM_Init: Failed to open handle to diag driver, error = 2`
`08-17 13:13:07.674 E/QCOMSysDaemon(11441):  Diag_LSM_Init failed : 0`
`08-17 13:13:07.928 W/ActivityManager(  877): getTasks: caller 10079 does not hold GET_TASKS; limiting output`
`08-17 13:13:07.930 W/ActivityManager(  877): getRecentTasks: caller 10079 does not hold GET_TASKS; limiting output`
`08-17 13:13:08.124 I/MSM-irqbalance(  381): Decided to move IRQ130 from CPU4 to CPU7`

**Reply by ajwfrost** [reply]
Well we can see it's loading, taking a while doing things, and then there's a "signal 9" i.e. "kill". Not sure what might trigger that..! it's not long enough for an ANR I think..
```
08-17 13:13:04.152 I/AIR RUNTIME(11403): Version 33,1,1,575 - platform Android-ARM64
08-17 13:13:04.933 I/Choreographer(11403): Skipped 50 frames! The application may be doing too much work on its main thread.
08-17 13:13:05.055 I/Process (11403): Sending signal. PID: 11403 SIG: 9
08-17 13:13:05.167 I/WindowState( 877): WIN DEATH: Window{19cc2fe1 u0 air.com.mediapro.UISampler/air.com.mediapro.UISampler.AIRAppEntry}
```
Given that we appear to be killing ourselves (process 11403 killing PID 11403) and with the earlier message, it may be that something is spinning and either we've determined that there's a problem with CPU load or (more likely?) there's a recursive call or an out of memory situation.

Can we get the AAB file of this please, to do some testing? Or rather, if you build with "aab-debug" target type, we can then deploy it and replace the AIR runtime library with our own one to see what's happening...
https://transfer.harman.com/requests/Ge8L4VJcBF62uK4eOYTOgQ

thanks

   Andrew

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - I have uploaded the aab and apk files for your investigation.

Thanks.

**Reply by ajwfrost** [reply]
Thanks - so it looks like one ANE is working and the other isn't:
```
08-18 06:23:33.564 32642 32642 I air.com.mediapro.UISampler: function aneVersion() = 1
08-18 06:23:33.564 32642 32642 I air.com.mediapro.UISampler: function getTestString() = Hello World. This is your ANE talking.
08-18 06:23:33.565 32642 32642 I air.com.mediapro.UISampler: function aid() = 16ae0d961f41fb7c
08-18 06:23:33.569  5517  7867 E StorageManagerService: Failed to mkdirs for /storage/55CE-2461/Android/data/air.com.mediapro.UISampler/files/ , volume is not writable
08-18 06:23:33.569 32642 32642 I air.com.mediapro.UISampler: function getExternalFilesDir() = /storage/emulated/0/Android/data/air.com.mediapro.UISampler/files
08-18 06:23:33.571 32642 32642 I air.com.mediapro.UISampler: ArgumentError: Error #3500: The extension context does not have a method with the name hasInitialized.
08-18 06:23:33.571 32642 32642 I air.com.mediapro.UISampler: 	at flash.external::ExtensionContext/_call()
08-18 06:23:33.571 32642 32642 I air.com.mediapro.UISampler: 	at flash.external::ExtensionContext/call()
08-18 06:23:33.571 32642 32642 I air.com.mediapro.UISampler: 	at com.mediapro::SampleXtra/hasInitialized()
```

I think I can see why this is happening: the extension.xml file lists the native library as "libSampleXtra.so" but in the APKS file you can see in the base-arm64_v8a.apk file that the .so file is now called "com.mediapro.SampleXtra.so"

... and I've just checked in the ADT code and can see that we're changing the filename for some reason to reflect the ANE name. That may have been to avoid conflicts in the names but I think isn't going to work (and I can't see that there would be a difference between the potential for a conflict in how the APK packaging works..)

So -> another bug, sorry, we'll get this updated... although this time as a workaround, if you're able to re-package the ANE with the filename in your extension.xml file matching this ("com.mediapro.SampleXtra.so") then it should work..!

Just checking: when you click the X button in the top bar of your app, is this calling NativeApplication.exit(), or something different? We're seeing the sigkill signal being raised when hitting this...

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - Unfortunately, I observe the same behavior as before. The native ANE does not work, even after updating the extension.xml file (I also had to update the .so name because ADT had complained about missing "native implementation" while packaging the ANE).

Again, the apk versions work perfectly. However, if you observe, ADT has saved the file as "libcom.mediapro.SampleXtra.so" in the "lib/arm64-v8a/"  folder. Still, the apk version installs and works as expected.

Therefore, I feel the workaround wouldn't work. You would probably have to update something at your end. If you open up a transfer request link, I will upload all these updated files for you to investigate further.

On a sidenote, all these issues are forcing me to explore Android (logcat, adb commands, etc.) in more detail. So, I have no complaints, as I am learning a lot.

Finally, when you click the X button, the app does the necessary cleanup and eventually calls `NativeApplication.exit()`.

Thanks.

**Reply by ajwfrost** [reply]
Hi @manishpatel0404  - thanks for checking that; but yes, we definitely need to update ADT for this. We've got an interim test build that I've just uploaded to the below link, based on 33.1.1.575 so if you can copy the adt.jar from here into the "lib" folder of that SDK, it should help..
https://transfer.harman.com/message/SrYtG16OOjQ0IZXF3u1KRI

I think we may check how that `NativeApplication.exit()` call is working then, it doesn't seem to be the most graceful of shutdowns, just calling 'kill' on itself! so at the very least we should close down the window rather than letting Android work out that the app has died..!

thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - Thanks for the updated adt.jar

I reset the filenames in the SampleXtra.ane to the original filenames and reset the extension.xml to its original form as well. When I output the APKS file, you can see in the base-arm64_v8a.apk file that the .so file is now "libSampleXtra.so", as expected.

However, when I install the APKS, the behavior has not changed. It still shows nothing for the SampleXtra output. Again, installing the apk version works perfectly.

I can upload the files for you to investigate further.

Thanks.

**Reply by ajwfrost** [reply]
Thanks @manishpatel0404  .. actually I should have looked in the APK file. So if you check in there, the library filename has been changed to "libcom.mediapro.SampleXtra.so"  ... so that's where it came from, the AAB generation was just missing off the "lib" at the start! I just checked the code that loads in the library from C++ and that does have essentially `"lib" + extensionID + ".so"` ...

Will get an updated version created for this..!
thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - You are absolutely correct. The apk file had 'always' stored the native lib as `"lib" + extensionID + ".so"`. Even I should have checked this, at least once you pointed out the issue in the APKS file. It would have saved us all a lot of time. Sorry about that. 

Borrowing a quote from a James Bond film, 
> Sometimes we are so focused on our enemies, we forget to watch our friends.

Looking forward to the update.

Thanks.

**Reply by ajwfrost** [reply]
Casino Royale? I like this from the sequel:
> When someone says 'we have people everywhere', you expect it to be hyperbole. Lots of people say that. Florists use that expression. It doesn't mean there's somebody working for them inside the bloody room!

Anyway ->  https://transfer.harman.com/message/OMsxzq2AtV1mJOC00oKDFj
should finally sort it out!

thanks

**Reply by manishpatel0404** [reply]
Ha, Ha, @ajwfrost - Good one. Yeah, I liked it too.

I will try out the update adt.jar file in a couple of hours and let you know how it went.

Thanks, once again.

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - When I run the app from the test device, I am now able to see the output of SampleXtra. The app installs perfectly using bundletool, as well as adt installApp (I know it calls bundletool internally, but I checked this, as well).

I was about to report this back to you when I decided to try the same with one of the apps we are trying to put out on the play store. The old behavior of the app starting and being killed immediately, happened again.

This is a K-12 learning app for both, Windows and Android platforms with some minor UI adaptations for the smaller Android screen. The AS3 code imports both Windows/Android ANEs and conditionally uses the appropriate one, based on the platform. We found it convenient for both platforms to share a common application manifest. So naturally, we listed the Windows and Android ANEs together in the extensions node. This actually works when packaging an apk with ADT. So, apk->install->launch works as expected on a device.

However, with aab->apks->install->launch, the app crashed upon launch on the device. I decided to investigate the issue before crying wolf, just yet. Turns out, if I comment the Windows ANE from the list of extensions in the application manifest, the aab->apks->install->launch works perfectly on the device.

It seems that ADT is more tolerant when packaging an apk file, as opposed to an aab file.

The workaround is simple. If we have two different manifest files, one for Windows and another for Android, there problem goes away. The only difference in the two files being the commented line in the Android manifest.

It is not a huge problem, but I thought you should know.

Finally, a big thanks, once again for the update. I can now proceed with figuring out the play store process.

**Reply by marchbold** [reply]
Out of interest so you have a complete default implementation in your windows ane?

**Reply by manishpatel0404** [reply]
Hi @marchbold - Yes. the Windows ANE has a default implementation.

I get the following warning message, whether I am building an aab or an apk, using ADT.

`Warning: Using default implementation for native extension 'com.mediapro.WinSampleXtra'`

However, despite the warning, the apk seems to install and launch perfectly. Only the aab installs and crashes upon launch.

**Comment by ajwfrost** [comment, marked as answer]
Hi @manishpatel0404  - thanks for pointing this out. An interesting few findings:
1) the warning about default implementation doesn't always come out, it depended on the order in which the .ane files were read from the disk vs the order of the ANEs listed in the extension file! which also meant that some validation checks were being skipped sometimes...
2) the AAB packaging was only including ANEs that had an Android platform in their support. For ones where the default library was the only 'valid' platform, they were still being skipped... so we just need to add the extension XML and default SWF into the bundle, like it does for the APK packaging, and then it works again
3) where you were seeing it crash - actually it was aborting in the bootstrap code with a helpful error code and message that wasn't being output anywhere. We're now outputting any such error message into the log so that it's easier to find out what's going on!

thanks

**Reply by manishpatel0404** [reply]
Hi, @ajwfrost - Look forward to the update when it is ready and available. It turned out to be a great learning experience for me.

Thanks.

**Reply by ajwfrost** [reply]
Hi @manishpatel0404 
This is working for our test cases now, are you able to try it and confirm?
https://transfer.harman.com/message/Hn0uKXIqQQHARz1dRhERvH
thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - IT WORKS LIKE A CHARM.

I can see that `WinSampleXtra` is now included in `base-master.apk`. The apks gets installed on the test device and the app launch and behavior is perfect.

Thanks for all the effort you have put into this.
